### DISCUSSION TITLE
on pull_request, how can I get the changes between the last pushed commit and the current commit in the PR branch

### DISCUSSION THREAD

**Discussion by vicaba** [tj-actions/changed-files#1549] [Q&A] [answered]
Hello,

## Scenario

I have the following MWE:
```
name: Scala CI

on:
  push:
    branches: [ "main" ]
  pull_request:
    branches: [ "main" ]

permissions:
  contents: read

jobs:
  check-changes:
    runs-on: ubuntu-latest
    outputs:
      all-relevant-changed: ${{ steps.changed-files-yaml.outputs.all-relevant_any_changed }}
    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 2
      - name: Get all relevant files that have changed
        id: changed-files-yaml
        uses: tj-actions/changed-files@v39
        with:
          since-last-remote-commit: 'true'
          files_yaml: |
            all-relevant:
              - '**'
              - '!**.md'
  build:
    runs-on: ubuntu-latest
    needs: [check-changes]
    if: needs.check-changes.outputs.all-relevant-changed == 'true'

    steps:
      - uses: actions/checkout@v4

```

The scenario is in a PR to "main".

## Expected result:
When modifying only "README.md" in the PR branch I do not want to trigger the `build` job, so I am using this action to prevent that.

## Problem:
Pushing changes to "README.md" to the PR branch still triggers this CI.

## Explanation and question:
I think this is because the CI is triggered by the event **on pull request to main**. Debugging the workflow shows (well, that's what I think) that the action compares the files between `main` and the `PR` branch. So any previous change made to another file is taken into account. If I only listen to `on: push` events, this works as expected. Using `on: push` will trigger the CI on any push, even in branches that are not PRs to main. How can I only trigger the CI in PRs that target main, but compare pushed commit and preceding commit from the same PR?
I also understand that the behaviour I am seeing is the expected one: https://github.com/tj-actions/changed-files/discussions/1379.

Thank you,

**Comment by jackton1** [comment, marked as answer]
Hi @vicaba, the workflow is getting triggered due to the use of an invalid input. 

I assume you meant to use [since_last_remote_commit](https://github.com/tj-actions/changed-files/discussions/1549#input_since_last_remote_commit) which should be separated by underscores.

**Reply by vicaba** [reply]
Thanks. Changed. However, now I get this error:
```
fatal: Needed a single revision
  Error: Unable to locate the commit sha: 47568a35dbcb87640c05e44bb41a41749b6a5fcd
  Error: Please verify that the commit sha is correct, and increase the 'fetch_depth' input if needed
  ##[debug]fatal: Needed a single revision
  ##[debug]
  Warning: Unable to locate the remote branch head sha. Falling back to the previous commit in the local history.
  /usr/bin/git rev-list -n 1 HEAD^
  d74d8a4a6bd549153b4f79301345cc9b0b157991
  /usr/bin/git merge-base d74d8a4a6bd549153b4f79301345cc9b0b157991 2538619192f2e7c9e62ae6a85467311e048f1e47
  Warning: Unable to find merge base between d74d8a4a6bd549153b4f79301345cc9b0b157991 and 2538619192f2e7c9e62ae6a85467311e048f1e47
  /usr/bin/git rev-parse --verify d74d8a4a6bd549153b4f79301345cc9b0b157991^{commit}
  d74d8a4a6bd549153b4f79301345cc9b0b157991
  ##[debug]Previous SHA: d74d8a4a6bd549153b4f79301345cc9b0b157991
  /usr/bin/git diff --quiet d74d8a4a6bd549153b4f79301345cc9b0b157991 2538619192f2e7c9e62ae6a[854](https://github.com/vicaba/phr-hippo-api-http/actions/runs/6123845033/job/16622665856?pr=12#step:3:855)67311e048f1e47
  Warning: Error checking commit history
  Error: Unable to determine a difference between d74d8a4a6bd549153b4f79301345cc9b0b157991..2538619192f2e7c9e62ae6a85467311e048f1e47
  ##[debug]Node Action run completed with exit code 1
  ##[debug]Finishing: Get all relevant files that have changed
```
https://github.com/vicaba/phr-hippo-api-http/actions/runs/6123845033/job/16622665856?pr=12

I understand that it is this line here.
**Error: Unable to determine a difference between d74d8a4a6bd549153b4f79301345cc9b0b157991..2538619192f2e7c9e62ae6a85467311e048f1e47**

Commit d74d8a4a6bd549153b4f79301345cc9b0b157991 is in "main" while Commit 2538619192f2e7c9e62ae6a85467311e048f1e47 is in the PR branch (the most recent push).

More than 2 (`fetch-depth`) commits between both, so this is why it fails (I set `fetch-depth` to 2). However, as far as I understand you said, it should compare the two most recent commits in the PR, so `fetch-depth` 2 should be enough.

**Reply by jackton1** [reply]
Hi @vicaba It appears the checkout action is checking out a merge of the PR HEAD and the base branch and in this case the fetch-depth doesn't work as expected. See: https://github.com/actions/checkout/issues/881

<img width="1308" alt="Screenshot 2023-09-08 at 10 28 48 AM" src="https://github.com/tj-actions/changed-files/assets/17484350/bbae7686-d5b3-4bb6-bc01-9b54e661df38">

### To resolve it you'll need to use

```yml
      - uses: actions/checkout@v4
        with:
          fetch-depth: 2
          ref: ${{ github.event.pull_request.head.sha }}
```

Or alternatively use the `push` event.

**Reply by jackton1** [reply]
Replicated here: https://github.com/tj-actions/changed-files/actions/runs/6124798182/job/16625530043?pr=1551#step:4:50
Fix here: https://github.com/tj-actions/changed-files/actions/runs/6124927163/job/16625951157?pr=1551#step:4:50

**Reply by vicaba** [reply]
Great. You helped me a lot and a learnt to better debug and understand Github Actions. I came across the issue you pointed aut but I didn't understand it fully and thought it was not related.
Also, thank you for taking the time to reproduce.
